### PR TITLE
Add support for systems which are missing arc4random_buf

### PR DESCRIPTION
**Pull request by cellularmitosis**
~~This PR adds a trivial implementation of `arc4random_buf()` (based on https://stackoverflow.com/a/12956868/558735) for systems which have `arc4random()` but not `arc4random_buf()`.  This includes PowerPC Macs which are stuck on OS X Tiger.~~

~~This is enabled by uncommenting the `JANET_NO_ARC4RANDOM_BUF` define in `janetconf.h`.~~

This PR modifies `os_cryptorand()` to use `/dev/urandom` on Macs prior to OS X 10.7.

context: https://github.com/janet-lang/janet/issues/430

### CODE REVIEW

**Comment by cellularmitosis** [comment]
oops, I ended up with a conflict after migrating this code.

**Comment by bakpakin** [comment]
This implementation can overflow the stack if size is too large. As a matter fo style and correctness, we don't allocate variable sized arrays on the stack in Janet. For example, `(os/cryptorand 0x7FFFFF)` will probably cause a segfault. Instead, assign directly into the output buffer.

**Comment by bakpakin** [comment]
Also, `(nbytes+3)/4` is a better expression for the number of required arc4random() calls.

**Comment by cellularmitosis** [comment]
Great insights, thanks for taking a look!  I will revise.

**Comment by cellularmitosis** [comment]
How about something like this?

```c
void arc4random_buf(void *buf, size_t nbytes) {
    uint32_t *buf_as_words = (uint32_t*)buf;
    size_t nwords = nbytes / 4;
    for (size_t i=0; i < nwords; i++) {
        buf_as_words[i] = arc4random();
    }

    size_t tail_len = nbytes % 4;
    if (tail_len) {
        uint8_t *tail = buf + nbytes - tail_len;
        uint32_t rand = arc4random();
        memcpy(tail, &rand, tail_len);
    }
}
```

**Comment by andrewchambers** [comment]
I don't know if this is a way to name things, obviously windows has no arc4random_buf but this implementation is not valid for windows.

**Comment by andrewchambers** [comment]
Is there no way to simply check for the maxos version? According to your comment you just need to check if the version is greater than 10.7

File: src/core/os.c [hunk 1]
```diff
@@ -70,6 +70,20 @@ extern char **environ;
 void arc4random_buf(void *buf, size_t nbytes);
 #endif
 
+/* arc4random_buf wasn't available in OS X until 10.7. */
+#ifdef JANET_NO_ARC4RANDOM_BUF
+/* Based on https://stackoverflow.com/a/12956868/558735 */
+uint32_t arc4random(void);
+void arc4random_buf(void *buf, size_t nbytes) {
```
Comment (andrewchambers) [inline src/core/os.c:77]: I guess you want this to be static.

**Comment by cellularmitosis** [comment]
@andrewchambers  Thanks for taking a look!

> I don't know if this is a good way to name things

I think I misunderstand.  If my system were missing `memcpy()`, and I wanted to provide a fill, wouldn't I be forced to name it `memcpy()`?

> Is there no way to simply check for the macos version

From the command line, yes.  We could check the version of the Darwin kernel via `uname -r`.  Would we add a line to the `Makefile` which would call `uname` and then automatically append `#define JANET_NO_ARC4RANDOM_BUF` to `janetconf.h`?

Or do you mean import some header file from which we can determine the kernel version?

**Comment by cellularmitosis** [comment]
@andrewchambers taking another look at `os_cryptorand()`, the Linux implementation uses `/dev/urandom`, and I just verified that device is also available on Tiger:

```
cell@emac$ uname -a
Darwin emac.home 8.11.0 Darwin Kernel Version 8.11.0: Wed Oct 10 18:26:00 PDT 2007; root:xnu-792.24.17~1/RELEASE_PPC Power Macintosh powerpc

cell@emac$ ls -l /dev/urandom 
crw-rw-rw-   1 root  wheel    8,   1 Dec 31  1969 /dev/urandom
```

So that's another route we could take.

Perhaps `#elif defined(JANET_LINUX)` would become something like `#elif defined(JANET_LINUX) && <thing which checks for OS X < 10.7>`

**Comment by cellularmitosis** [comment]
Ok, it looks like we could use `/usr/include/AvailabilityMacros.h`, which defines a range of OS X versions.

On Tiger, I see:

```
#define MAC_OS_X_VERSION_10_0 1000
#define MAC_OS_X_VERSION_10_1 1010
#define MAC_OS_X_VERSION_10_2 1020
#define MAC_OS_X_VERSION_10_3 1030
#define MAC_OS_X_VERSION_10_4 1040
```

Notably, these definitions stop at `10_4`.

I just checked my Mojave box and verified that the definitions stop with `10_14_6`:

```
#define MAC_OS_X_VERSION_10_14      101400
#define MAC_OS_X_VERSION_10_14_1    101401
#define MAC_OS_X_VERSION_10_14_4    101404
#define MAC_OS_X_VERSION_10_14_6    101406
```

So maybe we could use `!defined(MAC_OS_X_VERSION_10_7)`.

I guess we would also need to guard the include statement:

```
#ifdef __APPLE__
#include <AvailabilityMacros.h>
#endif
```

Using `/dev/urandom` with the above macro approach might be considered cleaner than bodging in an `arc4random_buf` fill.

**Comment by cellularmitosis** [comment]
just performed a quick local hack to confirm that the `/dev/urandom` approach works as expected under Tiger:

```
cell@emac$ ./build/janet
Janet 1.10.1-local  Copyright (C) 2017-2020 Calvin Rose
janet:1:> (os/cryptorand 1)
@"<"
janet:2:> (os/cryptorand 2)
@"\xAE\x19"
janet:3:> (os/cryptorand 3)
@"s\x15>"
janet:4:> (os/cryptorand 4)
@"L%\xAAR"
```

**Comment by andrewchambers** [comment]
To me either way seems ok to me if we check ( defined(JANET_APPLE) && !defined(MAC_OS_X_VERSION_10_7) ) like you said.

**Comment by andrewchambers** [comment]
looks good to me.

**Comment by cellularmitosis** [comment]
Ok, this is ready for a re-review now.  Thank you for obliging this PowerPC Mac rabbit hole :)

**Comment by bakpakin** [comment]
Looks alright to me - I actually prefer the arc4random code, as it will still work properly when running from within a chroot, but the feature detection looks good. And it is less code. Merging, thanks for the work here.

**Pull request merged**